Print pointer variables correctly in the scheduler log

iawia002 opened the pull request:
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

https://github.com/karmada-io/karmada/blob/85920e7ef49afc66325b85a37de97a662a7913c1/pkg/scheduler/core/generic_scheduler.go#L62

Currently we printed the `feasibleClusters` information in the generic_scheduler, but the `feasibleClusters` is an array of pointers, which makes the printed log completely unreadable:

```
I1206 06:22:54.370778       1 generic_scheduler.go:62] feasible clusters found: [0xc000858000 0xc000858200]
```

So I added a `String()` method to the `Cluster` object to make the log readable:

```
I1206 06:51:59.916405       1 generic_scheduler.go:62] feasible clusters found: [member1 member2]
```

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required.
-->
```release-note
NONE
```

/assign @RainbowMango

gy95 commented:
will not this patch  also print like this?
`I1206 06:22:54.370778       1 generic_scheduler.go:62] feasible clusters found: [0xc000858000 0xc000858200]`

iawia002 commented:
> will not this patch also print like this? `I1206 06:22:54.370778 1 generic_scheduler.go:62] feasible clusters found: [0xc000858000 0xc000858200]`

Nope, as I mentioned in the description, after this patch, it will print like this:

```
I1206 06:51:59.916405       1 generic_scheduler.go:62] feasible clusters found: [member1 member2]
```

And we can customize the information we want to print by updating the content returned by the `String` method.

gy95 commented:
Is this function that is implemented by package `klog`?
I use `fmt` for an example, but it still print the address, but not the struct infos.
```
package main

import (
	"fmt"
)

type cluster struct {
	name string
}

func main() {
	slice := []*cluster{
		&cluster{
			name: "a",
		},
		&cluster{
			name: "b",
		},
	}
	fmt.Printf("%v", slice)
}

func (c *cluster) string() string {
	return c.name
}
```

iawia002 commented:
> Is this function that is implemented by package `klog`? I use `fmt` for an example, but it still print the address, but not the struct infos.
> 
> ```
> package main
> 
> import (
> 	"fmt"
> )
> 
> type cluster struct {
> 	name string
> }
> 
> func main() {
> 	slice := []*cluster{
> 		&cluster{
> 			name: "a",
> 		},
> 		&cluster{
> 			name: "b",
> 		},
> 	}
> 	fmt.Printf("%v", slice)
> }
> 
> func (c *cluster) string() string {
> 	return c.name
> }
> ```

The `String` method needs to be exported:

```diff
-func (c *cluster) string() string {
+func (c *cluster) String() string {
```

gy95 commented:
I got this usage for the first time, I thought it was only for `%s` if we define the `String()` function 😂

RainbowMango commented:
I was thinking to add some methods for our APIs(structs), just you did in this PR. 
But I don't know if this is a good practice, do you know any examples from other projects?

iawia002 commented:
> I was thinking to add some methods for our APIs(structs), just you did in this PR. But I don't know if this is a good practice, do you know any examples from other projects?

I know that chaos-mesh is doing this, eg: https://github.com/chaos-mesh/chaos-mesh/blob/master/api/v1alpha1/awschaos_types.go#L101

I think we can add some public util functions when we encounter specific scenarios in the future (such as this one now).

RainbowMango commented:
Thanks. All the methods added to the APIs should be considered as a part of the API, which means we need to deal with caution.
When we make changes to the methods should follow consistent compatibility guarantees.

RainbowMango commented:
There might be quite a lot of methods in the future, do you think we should separate them from the `*_types.go`? 
I mean to introduce a new file under the same package.

iawia002 commented:
> There might be quite a lot of methods in the future, do you think we should separate them from the `*_types.go`? I mean to introduce a new file under the same package.

Yes, for better readability, we should separate them, but what should the file name be, `*_utils.go`?

RainbowMango commented:
> we should separate them, but what should the file name be, *_utils.go?

That's an option. No preference yet.

iawia002 commented:
> > we should separate them, but what should the file name be, *_utils.go?
> 
> That's an option. No preference yet.

I'm using this for now, we can change it if we have a better idea.

RainbowMango commented:
OK to me. But should we use `cluster_utils.go`? We can follow the naming rule in the other packages(e.g. policy) too.

And I guess a test file is necessary. We can test it by formats:
| format | receiver | output |
| --- | --- | --- |
| %s | variable | |
| %s | pointer | |
| %v | variable | |
| %v | pointer | |

iawia002 commented:
> OK to me. But should we use cluster_utils.go? We can follow the naming rule in the other packages(e.g. policy) too.

Reasonable, I have updated.

> And I guess a test file is necessary.

I have added some tests, but `%s variable` is not testable because the `%s` placeholder does not match the type `Cluster`.

RainbowMango commented:
/lgtm
/approve
